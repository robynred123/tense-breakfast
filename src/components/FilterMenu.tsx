import { useState, useEffect } from 'react';
import {
  Box,
  Checkbox,
  FormControlLabel,
  FormGroup,
  Grid,
  TextField,
  Typography,
} from '@mui/material';
import { DatePicker, LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';

import { useAppDispatch } from '../store';
import { DARK_GREY, GREY } from '../constants/colours';
import { AppointmentType, FilterOptions, Specialisms, DateRange } from '../types';
import { ButtonComponent } from './Button';
import { updateFilterOptions } from '../actions/actions';
import { updateSpecialismsArray, updateTypesArray } from '../utils/filterOptions';

interface MenuProps {
  filterOptions: FilterOptions;
  therapistSpecialisms: Specialisms[];
}

type Type = 'date' | 'type' | 'specialism';

export const FilterMenu = (props: MenuProps) => {
  const { filterOptions, therapistSpecialisms } = props;
  const { appointmentType, dateRange, specialisms } = filterOptions;
  const [startDate, setStartDate] = useState<Date | null>(null);
  const [endDate, setEndDate] = useState<Date | null>(null);
  const dispatch = useAppDispatch();

  // TODO: Move this to a util file
  const handleFilterChanges = (value: AppointmentType | Specialisms | DateRange, type: Type) => {
    const newOptions: FilterOptions = {
      appointmentType: appointmentType,
      dateRange: {
        start: dateRange.start,
        end: dateRange.end,
      },
      specialisms: specialisms,
    };
    switch (type) {
      case 'type':
        return dispatch(
          updateFilterOptions({
            ...newOptions,
            appointmentType: updateTypesArray(value as AppointmentType, appointmentType),
          })
        );
      case 'specialism':
        return dispatch(
          updateFilterOptions({
            ...newOptions,
            specialisms: updateSpecialismsArray(value as Specialisms, specialisms),
          })
        );
      case 'date':
        return dispatch(
          updateFilterOptions({
            ...newOptions,
            dateRange: value as DateRange,
          })
        );
      default:
        break;
    }
  };

  // only filter when both dates have been selected - reducers number of actions dispatched & reloads
  useEffect(() => {
    if (startDate && endDate) {
      const dateRange = {
        start: startDate.toDateString(),
        end: endDate.toDateString(),
      };
      handleFilterChanges(dateRange, 'date');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [startDate, endDate]);

  const determineSelectedType = (value: AppointmentType) => {
    if (appointmentType.includes(value)) {
      return 'green';
    } else return 'grey';
  };

  return (
    <Box
      sx={{
        flexGrow: 1,
        width: 1,
        flexDirection: 'column',
        display: 'flex',
        backgroundColor: GREY,
        border: 'double',
        borderColor: DARK_GREY,
        borderRadius: 10,
        justifyContent: 'center',
        alignContent: 'center',
      }}
    >
      <Grid container sx={{ alignContent: 'center', flexDirection: 'column' }}>
        <Grid item m={2}>
          <Typography variant='h4' fontFamily={'lato, sans-serif'}>
            Filter Results
          </Typography>
        </Grid>
      </Grid>
      <Grid container sx={{ flexDirection: 'column', padding: '0 2em 2em 2em' }}>
        <Grid container>
          <Grid item>
            <Typography variant='h5' fontFamily={'lato, sans-serif'}>
              Dates
            </Typography>
          </Grid>
          <Grid container sx={{ width: '100%', justifyContent: 'center' }}>
            <FormGroup>
              <LocalizationProvider dateAdapter={AdapterDateFns}>
                <DatePicker
                  label='Start Date'
                  value={startDate}
                  onChange={(newValue) => {
                    setStartDate(newValue);
                  }}
                  minDate={new Date()}
                  renderInput={(params) => <TextField {...params} sx={{ margin: '1em' }} />}
                />
                <DatePicker
                  label='End Date'
                  value={endDate}
                  onChange={(newValue) => {
                    setEndDate(newValue);
                  }}
                  minDate={startDate || new Date()}
                  renderInput={(params) => <TextField {...params} sx={{ margin: '1em' }} />}
                />
              </LocalizationProvider>
            </FormGroup>
          </Grid>
        </Grid>
        <Grid container sx={{ marginBottom: '2em' }}>
          <Grid item>
            <Typography variant='h5' fontFamily={'lato, sans-serif'}>
              Appointment Type
            </Typography>
          </Grid>
          <Grid container sx={{ width: '100%', justifyContent: 'space-evenly' }}>
            <ButtonComponent
              onClick={() => handleFilterChanges('one_off', 'type')}
              text='One Off'
              buttonColour={determineSelectedType('one_off')}
              disabled={false}
              width={'40%'}
            />
            <ButtonComponent
              onClick={() => handleFilterChanges('consultation', 'type')}
              text='Consultation'
              buttonColour={determineSelectedType('consultation')}
              disabled={false}
              width={'40%'}
            />
          </Grid>
        </Grid>
        <Grid container>
          <Grid item>
            <Typography variant='h5' fontFamily={'lato, sans-serif'}>
              Specialism
            </Typography>
          </Grid>
          <Grid container sx={{ width: '100%', justifyContent: 'flex-start' }}>
            <FormGroup>
              {therapistSpecialisms &&
                therapistSpecialisms.map((specialism) => (
                  <FormControlLabel
                    key={specialism}
                    control={
                      <Checkbox
                        defaultChecked={false}
                        onChange={() => handleFilterChanges(specialism, 'specialism')}
                      />
                    }
                    label={specialism}
                  />
                ))}
            </FormGroup>
          </Grid>
        </Grid>
      </Grid>
    </Box>
  );
};
