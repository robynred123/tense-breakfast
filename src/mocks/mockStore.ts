import configureStore from 'redux-mock-store';
import { Store, AnyAction } from 'redux';
import { FilterOptions, ReducerState } from '../types/index';

export const filterOptions: FilterOptions = {
  appointmentType: [],
  dateRange: {
    start: null,
    end: null,
  },
  specialisms: [],
};
export const initialState: ReducerState = {
  therapists: [],
  filteredTherapists: [],
  filterOptions,
  mobileFilter: false,
  error: null,
};

export const mockStore = (initialState: ReducerState) => {
  const mockStore = configureStore();
  let store: Store<any, AnyAction>;
  store = mockStore({ one: initialState });
  return store;
};
