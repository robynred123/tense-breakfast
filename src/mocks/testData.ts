import { TherapistInfo } from '../types';

export const therapistOne: TherapistInfo = {
  id: '063eb398-45f7-4a9c-a5d3-e95d14c05c44',
  firstName: 'Albert',
  lastName: 'Einstein',
  appointment_types: ['one_off'],
  appointment_mediums: ['phone'],
  specialisms: ['Alcoholism', 'Gender identity', 'Gambling Addiction'],
};

export const therapistTwo: TherapistInfo = {
  id: '154aa58b-4c25-4d64-a7a2-9288a7d94468',
  firstName: 'Jackie',
  lastName: 'Chan',
  appointment_types: ['consultation', 'one_off'],
  appointment_mediums: ['video'],
  specialisms: ['Gender identity', 'Alcoholism'],
};
