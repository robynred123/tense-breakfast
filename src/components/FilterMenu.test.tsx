/* eslint-disable testing-library/await-async-query */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { mockStore, initialState, filterOptions } from '../mocks/mockStore';
import { FilterMenu } from './FilterMenu';

describe('<FilterMenu />', () => {
  const therapistSpecialisms = ['ADHD'];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Should render the FilterMenu component successfully', () => {
    const store = mockStore(initialState);
    const view = renderer.create(
      <Provider store={store}>
        <FilterMenu filterOptions={filterOptions} therapistSpecialisms={therapistSpecialisms} />
      </Provider>
    );
    expect(view.toJSON()).toMatchSnapshot();
    expect(view.root.findByType('h4').props.children).toEqual('Filter Results');
    expect(view.root.findAllByType('h5')[0].props.children).toEqual('Dates');
    expect(view.root.findAllByType('label')[0].props.children[0]).toEqual('Start Date');
    expect(view.root.findAllByType('label')[1].props.children[0]).toEqual('End Date');

    expect(view.root.findAllByType('h5')[1].props.children).toEqual('Appointment Type');
    const buttons = view.root.findAllByType('button');
    expect(buttons[0].props.children).toEqual([[undefined, 'One Off', undefined], null]);
    expect(buttons[1].props.children).toEqual([[undefined, 'Consultation', undefined], null]);

    expect(view.root.findAllByType('h5')[2].props.children).toEqual('Specialism');
    expect(view.root.findAllByType('input')[2].props.type).toEqual('checkbox');
    expect(view.root.findAllByType('input')[2].props.checked).toEqual(false);
    expect(view.root.findAllByType('label')[2].props.children[1].props.children).toEqual('ADHD');
  });
});
