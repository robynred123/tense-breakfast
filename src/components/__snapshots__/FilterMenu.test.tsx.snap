// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterMenu /> Should render the FilterMenu component successfully 1`] = `
<div
  className="MuiBox-root css-nr4a5l"
>
  <div
    className="MuiGrid-root MuiGrid-container css-1t09gft-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-1llorz5-MuiGrid-root"
    >
      <h5
        className="MuiTypography-root MuiTypography-h5 css-1mzsfvt-MuiTypography-root"
      >
        Filter Results
      </h5>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container css-14ldd2u-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 css-k0cgwl-MuiTypography-root"
        >
          Appointment Type
        </h6>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container css-p54wsc-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rw4q7k-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          One-Off
        </button>
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rw4q7k-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Consultation
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FilterMenu /> Should render the button component successfully 1`] = `
<div
  className="MuiBox-root css-nr4a5l"
>
  <div
    className="MuiGrid-root MuiGrid-container css-1t09gft-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-1llorz5-MuiGrid-root"
    >
      <h5
        className="MuiTypography-root MuiTypography-h5 css-1mzsfvt-MuiTypography-root"
      >
        Filter Results
      </h5>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container css-14ldd2u-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 css-k0cgwl-MuiTypography-root"
        >
          Appointment Type
        </h6>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container css-p54wsc-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rw4q7k-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          One-Off
        </button>
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-rw4q7k-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Consultation
        </button>
      </div>
    </div>
  </div>
</div>
`;
